### ISSUE TITLE
unexpected complaint 'unsupported-delete-operation'  on an object of type 'dict[str, str]'

### ISSUE THREAD

**Issue by crimson-gao** [PyCQA/pylint#5413]
### Bug description

unexpected complaint 'unsupported-delete-operation'  on an object of type 'dict[str, str]'
```python
p, r = conv_variables(dashboard["templating"]["list"])
  if p is not None:
      for key in pipeline_remove_keys:
          if p.__contains__(key):
              del p[key]     # where pylint complains
      res["pipeline"] = p
```
I'm using pylint of version 2.12.1 with default configuration.  
Here, p is an object of type dict[str, str].
![image](https://user-images.githubusercontent.com/39024757/143735088-5cfdb140-b943-40ac-968b-9619048f9924.png)

### Configuration

```ini
default
```

### Command used

```shell
python -m pylint dashboard.py
```

### Pylint output

```shell
dashboard.py:54:24: E1138: 'p' does not support item deletion (unsupported-delete-operation)
```

### Expected behavior

no complaint

### Pylint version

```shell
pyliny==2.12.1
astroid==2.9.0
```

### OS / Environment

Linux VM-162-135-centos 5.4.119-1-tlinux4-0005 #1 SMP Thu Jul 8 11:20:55 CST 2021 x86_64 x86_64 x86_64 GNU/Linux

**Comment by jacobtylerwalls** [comment]
Hi @crimson-gao, we will need just a little bit more information to work on this. A generic dict with strings as keys and values works fine, so something is up with `conv_variables(dashboard["templating"]["list"])`. That's probably the root of the problem; `pylance` (in your screenshot) can tell that it's `dict[str, str]`, but `pylint`/`astroid` can't.

Can you provide an example that we can reproduce from scratch? We need to know what packages you're relying on to create `p` so we can create `p` ourselves and look at the failure. Thanks.

generic example:
```shell
$ cat z.py
p = {'apple': 'banana'}
del p['apple']
$ pylint z  --disable=missing-module-docstring

--------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 10.00/10, +0.00)
```

**Comment by jacobtylerwalls** [comment]
Feel free to reopen if you can provide enough information for us to reproduce the issue.
